BuildLayers Task fails when jars have the same name

### Expected Behavior

The BuildLayers task should either:

- Gracefully handle duplicate entries using the default (INHERIT) strategy
- Allow the configuration of a [DuplicateStrategy](https://docs.gradle.org/7.5.1/javadoc/org/gradle/api/file/DuplicatesStrategy.html) in the task (or somewhere)

### Actual Behaviour

```
> Task :buildLayers FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':buildLayers'.
> Entry shared-1.0.0-SNAPSHOT.jar is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.5.1/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.

```

### Steps To Reproduce

1. Create two projects with a 'src' (main) module and a 'shared' module
2. In one of them, depend on both `shared` modules
3. Attempt to run a build

### Environment Information

Mac OSX
Java 17

### Example Application

_No response_

### Version

3.6.3

--- a/.github/workflows/graalvm.yml
+++ b/.github/workflows/graalvm.yml
@@ -20,7 +20,7 @@ jobs:
     strategy:
       matrix:
         java: ['11']
-        graalvm: ['latest', 'dev']
+        graalvm: ['latest']
     steps:
        # https://github.com/actions/virtual-environments/issues/709
       - name: Free disk space

--- a/docker-plugin/src/main/java/io/micronaut/gradle/docker/tasks/BuildLayersTask.java
+++ b/docker-plugin/src/main/java/io/micronaut/gradle/docker/tasks/BuildLayersTask.java
@@ -3,13 +3,18 @@ package io.micronaut.gradle.docker.tasks;
 import io.micronaut.gradle.docker.model.Layer;
 import io.micronaut.gradle.docker.model.LayerKind;
 import org.gradle.api.DefaultTask;
+import org.gradle.api.file.CopySpec;
 import org.gradle.api.file.Directory;
 import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.file.DuplicatesStrategy;
 import org.gradle.api.internal.file.FileOperations;
 import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.Input;
 import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.Optional;
 import org.gradle.api.tasks.OutputDirectory;
 import org.gradle.api.tasks.TaskAction;
 
@@ -20,6 +25,10 @@ public abstract class BuildLayersTask extends DefaultTask {
 
     @Nested
     public abstract ListProperty<Layer> getLayers();
+
+    @Input
+    @Optional
+    public abstract Property<DuplicatesStrategy> getDuplicatesStrategy();
 
     @OutputDirectory
     public abstract DirectoryProperty getOutputDir();
@@ -42,10 +51,22 @@ public abstract class BuildLayersTask extends DefaultTask {
             final Provider<Directory> layerDir = layerDirectoryOf(layer, getOutputDir(), libsDir, resourcesDir, classesDir);
             if (layer.getLayerKind().get() == LayerKind.APP) {
                 // special case for now
-                fileOperations.copy(copy -> copy.from(layer.getFiles()).into(getOutputDir()).rename(s -> "application.jar"));
+                fileOperations.copy(copy -> {
+                    configureDuplicatesStrategy(copy);
+                    copy.from(layer.getFiles()).into(getOutputDir()).rename(s -> "application.jar");
+                });
             } else {
-                fileOperations.copy(copy -> copy.from(layer.getFiles()).into(layerDir));
+                fileOperations.copy(copy -> {
+                    configureDuplicatesStrategy(copy);
+                    copy.from(layer.getFiles()).into(layerDir);
+                });
             }
+        }
+    }
+
+    private void configureDuplicatesStrategy(CopySpec copy) {
+        if (getDuplicatesStrategy().isPresent()) {
+            copy.setDuplicatesStrategy(getDuplicatesStrategy().get());
         }
     }
 

--- a/docker-plugin/src/test/groovy/io/micronaut/gradle/BuildLayersSpec.groovy
+++ b/docker-plugin/src/test/groovy/io/micronaut/gradle/BuildLayersSpec.groovy
@@ -41,4 +41,56 @@ class BuildLayersSpec extends AbstractGradleBuildSpec {
         new File(testProjectDir.root, "build-custom/docker/main/layers").exists()
         !new File(testProjectDir.root, "build/docker/main/layers").exists()
     }
+
+    void 'test build layers with duplicates strategy'() {
+        given:
+        settingsFile << "rootProject.name = 'hello-world'"
+        buildFile << """
+            plugins {
+                id "io.micronaut.minimal.application"
+                id "io.micronaut.docker"
+            }
+            
+            micronaut {
+                version "3.5.1"
+                runtime "netty"
+                testRuntime "junit5"
+            }
+            
+            $repositoriesBlock
+            
+            dependencies {
+                runtimeOnly("ch.qos.logback:logback-classic")
+                testImplementation("io.micronaut:micronaut-http-client")
+            }
+            mainClassName="example.Application"
+
+            def jar1 = tasks.register("otherJar", Jar) {
+                destinationDirectory = file("build/libs1")
+                archiveBaseName = 'toto'
+                from sourceSets.main.output
+            }
+            def jar2 = tasks.register("otherJar2", Jar) {
+                destinationDirectory = file("build/libs2")
+                archiveBaseName = 'toto'
+                from sourceSets.main.output
+            }
+
+            configurations.runtimeClasspath.dependencies.add(dependencies.create(files(jar1, jar2)))            
+
+            tasks.withType(io.micronaut.gradle.docker.tasks.BuildLayersTask) {
+                duplicatesStrategy = DuplicatesStrategy.INCLUDE
+            }
+
+        """
+
+        when:
+        def result = build('buildLayers')
+
+        def task = result.task(":buildLayers")
+
+        then:
+        task.outcome == TaskOutcome.SUCCESS
+        new File(testProjectDir.root, "build/docker/main/layers").exists()
+    }
 }

--- a/src/docs/asciidoc/index.adoc
+++ b/src/docs/asciidoc/index.adoc
@@ -655,6 +655,8 @@ If you wish to split your native image tests from your regular tests you can {gr
 
 === Docker Support
 
+==== Building docker images
+
 The Micronaut plugin includes integration with the https://bmuschko.github.io/gradle-docker-plugin[Gradle Docker plugin] allowing you to easily build applications and native images using Docker containers.
 
 Applications are built as layered JARs using the `buildLayers` task ensuring optimized Docker images for Java applications.
@@ -743,6 +745,8 @@ tasks.named<DockerBuildImage>("dockerBuildCrac") {
 
 Notice that you can supply two different image names to push to for the JVM version and the native version of the application.
 
+==== Customized docker files
+
 If you wish to customize the docker builds that are used, the easiest way is to run `./gradlew dockerfile` (or `dockerfileNative` for the native version) and copy the generated `Dockerfile` from `build/docker` to your root directory and modify as required.
 
 If you wish to customize the JVM arguments or native image arguments then it is possible to do so with the `args` method of the `dockerfile` and `dockerfileNative` tasks:
@@ -768,6 +772,8 @@ tasks.named<io.micronaut.gradle.docker.NativeImageDockerfile>("dockerfileNative"
 ----
 
 The above configuration uses a max heap setting of `128m` for Java and `64m` for native image for the application.
+
+==== Adding additional instructions
 
 To add additional docker instructions to the generated Dockerfile, such as adding a HEALTHCHECK, you can do the following. The additional instructions will be added at the end of the `Dockerfile` just before the `ENTRYPOINT`.
 
@@ -796,6 +802,27 @@ tasks.named<io.micronaut.gradle.docker.NativeImageDockerfile>("dockerfileNative"
 ----
 
 You can also add any of the other instructions/commands that the docker plugin supports, see {docker-plugin}/blob/master/src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy[the Dockerfile task documentation].
+
+==== Duplicates on classpath
+
+In some projects, it may happen that different transitive dependencies have the same file name.
+In this case, the task which builds the layers will fail with a duplicate error.
+You can workaround this issue by configuring the duplicates strategy on the task:
+
+[source, groovy, subs="verbatim,attributes", role="multi-language-sample"]
+----
+tasks.withType<io.micronaut.gradle.docker.tasks.BuildLayersTask> {
+    duplicatesStrategy = DuplicatesStrategy.INCLUDE
+}
+----
+
+[source, kotlin, subs="verbatim,attributes", role="multi-language-sample"]
+----
+tasks.withType(io.micronaut.gradle.docker.tasks.BuildLayersTask) {
+    duplicatesStrategy.set(DuplicatesStrategy.INCLUDE)
+}
+----
+
 
 === Micronaut Runtimes
 

